messaging/broker.go

```diff
@@ -92,25 +126,67 @@ func (b *Broker) Open(path string, u *url.URL) error {
 	}
 	b.path = path
 
-	// Require a non-blank connection address.
-	if u == nil {
-		return ErrConnectionAddressRequired
+	// Ensure root directory exists.
+	if err := os.MkdirAll(path, 0700); err != nil {
+		b.close()
+		return fmt.Errorf("mkdir: %s", err)
+	}
+
+	// Open meta file.
+	meta, err := bolt.Open(b.metaPath(), 0600, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		b.close()
+		return fmt.Errorf("open meta: %s", err)
 	}
+	b.meta = meta
 
-	// Read meta data from snapshot.
-	if err := b.load(); err != nil {
+	// Initialize data from meta store.
+	if err := b.meta.Update(func(tx *bolt.Tx) error {
+		tx.CreateBucketIfNotExists([]byte("meta"))
+
+		// Read in index from meta store, if set.
+		if v := tx.Bucket([]byte("meta")).Get([]byte("index")); v != nil {
+			b.index = btou64(v)
+		}
+
+		return nil
+	}); err != nil {
 		_ = b.close()
 		return err
 	}
 
-	// Open underlying raft log.
-	if err := b.log.Open(filepath.Join(path, "raft")); err != nil {
-		return fmt.Errorf("raft: %s", err)
+	// Read all topic metadata into memory.
+	if err := b.openTopics(); err != nil {
+		b.close()
+		return fmt.Errorf("open topics: %s", err)
+	}
+
+	return nil
+}
+
+// loadTopics reads all topic metadata into memory.
+func (b *Broker) openTopics() error {
+	// Read all topics from the broker directory.
+	topics, err := ReadTopics(b.path)
+	if err != nil {
+		return fmt.Errorf("read topics: %s", err)
 	}
 
-	// Copy connection URL.
-	b.log.URL = &url.URL{}
-	*b.log.URL = *u
+	// Open each topic and append to the map.
+	b.topics = make(map[uint64]*Topic)
+	for _, t := range topics {
+		if err := t.Open(); err != nil {
+			return fmt.Errorf("open topic: id=%d, err=%s", t.id, err)
+		}
+		b.topics[t.id] = t
+	}
+
+	// Retrieve the highest index across all topics.
```

Comment:
Yes, but I feel like it's clearer to load the topics in one section of the code and then determine highest index in another section.